### DISCUSSION TITLE
Switching UPDI and serial

### DISCUSSION THREAD

**Discussion by jblang** [SpenceKonde/DxCore#514] [Q&A] [answered]
Would it be possible to use the same serial adapter for both programming and communication?  I am thinking of using one or more mosfets with their gates driven low by DTR to enable UPDI. When DTR is high, they would let normal TX/RX lines through.  I may try to breadboard something up, but I was curious whether it has been tried or if there's any reason it couldn't work?

**Comment by jblang** [comment]
For now I have decided against this in the board I'm designing. I will instead use a DPDT switch to select between programming and normal serial comms. I can program via the bootloader in serial comms mode to avoid having to manually toggle the switch constantly.  Here's the relevant portion of my schematic.

![image](https://github.com/SpenceKonde/DxCore/assets/947942/57c1e37c-0613-4677-8145-a8c2119b19a1)

I have posted [more details](https://github.com/SpenceKonde/DxCore/discussions/515) about the board I will be using this circuit on in show and tell, if you care to have a look. 

Thanks for paving the way for hobbyists using the Dx series! I have found so much useful information in your docs even though I'm not using the Arduino core.

**Reply by PeterCharlesJones** [reply]
I appreciate a bit late in the day for a reply and 'respect ' to Dr Azzy, what he says is gospel - he developed the Arduino wrappers after all.

I often use the UPDI pin for debug TTY output combined with SendOnlySoftwareSerial.h  as it saves a pin.  (For ATtiny1604, specify Arduino pin 11 - PA0).
If you use the diode modified serial card programmer, you can get your TTY output through the same serial card you use for programming. No need to do anything  - oh - apart from POR the AVR after programming so the pin goes back to being a GPIO.

If you have time (I did), you can build your HV programmer to give the target device a POW after programming.
To do this, you need your HV programmer to supply target AVR with power (and hence stop this for POR). But if your HV programmer is supplying power to the target AVR, you'll need to make sure that it can either supply enough power for any connected stuff hanging of VCC, or have all this disconnected whilst the programmer is attached. Mine measures if there is external power on the target and if there is, it refuses to fire up the HV source as there would be no possibility of a target device POR in this case.

I only looked at this post as I was trying so see if anyone has succeeded in sending TTY debug to the UPDI pin whilst it was configured as RESET - that would be nirvana, a reset pin AND TTY debug from the UPDI port.  I tested this scenario and found I was unable to get  the chip to drive TTY output through the UPDI pin whilst it was defined as RESET (There was some noise but no detectable characters were received). Worth a try though.

If you want code examples or circuit board schematics, please reply to this post.

**Comment by SpenceKonde** [comment, marked as answer]
I dunno about the MOSFETs for this - the thing to remember is that SerialUPDI at a hardware level is a hair's breadth from not working on the tinyAVRs because of how weak the pin driver on their UPDI pin is (since it's also the reset pin and also the pin that getsthe HV pulse, HV pulse pin has always had crap drivers, which they finally admitted were pretty worthless, when they finally added Reset as I/O to Dx with the double D's - they didn't even bother with pin drivers. (on a log scale, the strength of the output drivers is about halfway between the pullup and a normal pin driver). But on tinyAVRs, it can only drive the pin low with like half a milliamp or something stupid like that (see the datasheet). 

I've definitely swtiched between modes with a physical switch, and I've also done CH343 dual serial adapters with one dedicated to UPDI and the other to normal serial. I'd be selling them already were it not for some strange issues that have resisted proper understanding. There's a parameter you can pass to SerialUPDI I think to set up it's DTR and RTS pins to whatever you want them to be, too, that I added so I could potentially support an HVSP scheme with it.

**Comment by jblang** [comment]
I breadboarded the circuit with a physical switch and confirmed that it works, though had to move the series resistor on the TX line to the other side of the switch to avoid parasitically powering the AVR though the TX line in either mode.  Also I am using 1K, not 4.7K as the schematic reads.

**Reply by jblang** [reply]
I made a protoboard version of my switchable programmer, complete with a ridiculously oversized switch, the only dpdt I had on hand. I have affectionately dubbed it "big chonky" 😆. It conveniently fits on pins 15-20 of a breadboarded AVR128DB28. 

![image](https://github.com/SpenceKonde/DxCore/assets/947942/1df6d8ed-b938-4208-969d-2c053dded18f)

**Comment by tileiar** [comment]
Hi, I'm using the USB to Dual Serial Ports Chip [CH342](https://www.wch-ic.com/products/CH342.html) for my programmer. So there is no hassle switching between UPDI and Serial. This would be a different approach to adress the problem.

**Reply by SpenceKonde** [reply]
same
